### PR TITLE
Field Validation API Concepts

### PR DESCRIPTION
**Pull request by kevindelgado**
Docs for https://github.com/kubernetes/enhancements/issues/2885

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP] placeholder for field validation beta docs

**Comment by tengqm** [comment]
@kevindelgado I have proposed something in #32592, for the same topic. Would you like review #32592 and see if that covers all changes needed?

**Comment by kevindelgado** [comment]
> @kevindelgado I have proposed something in #32592, for the same topic. Would you like review #32592 and see if that covers all changes needed?

Thanks for the PR. We can can go ahead and merge that. I think the one other piece of documentation we want, is something for the [api-references](https://kubernetes.io/docs/reference/using-api/api-concepts/#make-a-dry-run-request) page similar to what we have for `--dry-run`. I will use this PR to do that, and should have something up for review EOD today.

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:671]:
```suggestion
The two situations this will occur are:
```

?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:673]:
```suggestion
1. The field is unrecognized because it is not in the resource's OpenAPI schema. (One
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:675]:
```suggestion
   fields via `x-kubernetes-preserve-unknown-fields`).
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:678]:
```suggestion
When you use HTTP verbs that can submit data (`POST`, `PUT`, and `PATCH`), field
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:685]:
```suggestion
field that is supposed to be an integer, but is submitted as a string), will always
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (tengqm) [inline content/en/docs/reference/using-api/api-concepts.md:704]: Unknown or duplicate fields are ignored in this level, right?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [reply content/en/docs/reference/using-api/api-concepts.md:671]: How about:
```suggestion
The two situations where this will occur are:
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (apelisse) [reply content/en/docs/reference/using-api/api-concepts.md:704]: Yes, they are dropped but the warning is added to the response.

**Comment by apelisse** [comment]
Technical LGTM

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:702]:
```suggestion
uses the HTTP response header `Warning:`; see
```

**Review by sftim — DISMISSED** [review]
Remember to remove `placeholder.md` from this PR.

**Comment by kevindelgado** [comment]
All feedback addressed, PTAL

**Comment by kevindelgado** [comment]
/hold

Due to an issue with validating CRD metadata, we have pulled Server-Side Field Validation's beta graduation from 1.24 and will try again in 1.25

https://github.com/kubernetes/kubernetes/issues/109215

**Comment by kevindelgado** [comment]
closing this, will open a separate PR when the time comes against dev-1.25

**Comment by kevindelgado** [comment]
Reopening this PR and basing it against dev-1.25 for the upcoming release. Nothing should have changed in the docs between 1.24 and 1.25 since we were only fixing bugs.

/assign @liggitt

File: content/en/docs/reference/using-api/api-concepts.md
Comment (liggitt) [inline content/en/docs/reference/using-api/api-concepts.md:718]:
```suggestion
server-side field validation when sending requests to a server with this feature enabled.
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:664]: (nit)
```suggestion
## Field validation
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:674]: I'd add a tooltip:
```suggestion
   exception to this is for {{< glossary_tooltip term_id="CustomResourceDefinition" text="CRDs" >}}
   that explicitly choose not to prune unknown
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:694]: At the HTTP level, we expect a title case string here. Is that right?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:709]:
```suggestion
detects any unknown or duplicate fields. The response message from the API server specifies all the
unknown or duplicate fields that the API server has detected.
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:715]:
```suggestion
The `kubectl` tool uses the `--validate` flag to set the level of field validation.
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:722]:
```suggestion
(equivalent to `ignore`). The default validation setting for kubectl is `--validate=true`,
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:686]: I'd write:
```suggestion
If you submit a request that specifies an unrecognized field, and that is also invalid for
a different reason (for example, the request provides a string value where the API expects
an integer), then the API server responds with a 400 Bad Request error response.

You always receive an error response in this case, no matter what field validation level you requested.
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:692]: How about:
```suggestion
Field validation is set by the optional `fieldValidation` query parameter. The three
values that you can provide for this parameter are:
```

I don't think we need to tell the reader that `Ignore`, `Warn` and `Strict` are strings; either the reader is amply familiar with strings in this context, enough to work it out for themselves, or they have a lot more background reading to do before it's useful to explain that detail.

For a client, these are opaque values, and anyway, substrings of URLs are always strings.

**Review by sftim — COMMENTED** [review]
Some feedback on style. A lot of it is only nits.

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done, and bumped the version

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done.

**Review by kevindelgado — COMMENTED** [review]
Appreciate the detailed feedback. All comments addressed, PTAL

**Comment by liggitt** [comment]
technical content lgtm

**Comment by kevindelgado** [comment]
Rebased on dev-1.25. Nothing has changed since @liggitt gave the LGTM

PTAL @sftim

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:668]: (nit / aside)
```suggestion
By default, the API server drops usually drops fields that it does not recognize
```
(is that correct)?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:669]: nit: AIUI  the message from the client can include any resource body; it doesn't have to be an object.

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:679]:
```suggestion
Provided that the `ServerSideFieldValidation` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled (disabled
```

**Review by sftim — COMMENTED** [review]
Thanks

/lgtm
/approve

It'd be handy to log an issue about the link to the feature gates page (or a fixup PR). These links help readers.

**Comment by sftim** [comment]
/hold
Feel free to merge this

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md:669]: Yea, I don't think there's anything here that is saying messages are exclusively objects, just that objects received from the clients will have their unknown fields dropped by the api server

**Comment by kevindelgado** [comment]
feedback/nits addressed, please tag it again @sftim

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:666]: On reflection, I think this should be for the heading “Setting the field validation level” and not for the 2nd-level field validation heading. Even with the feature gate off, there's some (actually quite a lot) of field validation going on.

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:667]: We should describe more about the other forms of field validation (not related to this feature being documented).

Something like:
```markdown
Kubernetes always validates the type of fields. For example, if a field in the API is defined
as a number, you cannot set that field to a text value. If a field is defined as an array of strings,
you can only provide an array. Some fields allow you to omit them, other fields are required.
Omitting a required field from an API request is an error.

If you make a request with an extra field, one that the cluster's control plane does not recognize,
then the behaviour of the API server is more complicated.
```

What do you think?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:669]:
```suggestion
from an input that it receives (for example, the JSON body of a `PUT` request).
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:671]: (nit)
```suggestion
There are two situations where the API server drops fields that you supplied in an HTTP request.
These situations are:
```

File: content/en/docs/reference/using-api/api-concepts.md
Comment (sftim) [inline content/en/docs/reference/using-api/api-concepts.md:686]: Should this come after the heading “Setting the field validation level”?

**Review by sftim — COMMENTED** [review]
One more thought. Assume that some folk don't read the “Setting the field validation level” text because they know it's not enabled for their cluster.

When someone supplies duplicate values and these are eventually accepted, it would be nice to clearly explain which value wins. For JSON, it's the _last_ duplicate - right?

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md:674]: yep, that sounds good to me. I agree it seems weird to jump right into the unknown field bits when there's more to validation than just that.

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: moved

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md]: done

File: content/en/docs/reference/using-api/api-concepts.md
Comment (kevindelgado) [reply content/en/docs/reference/using-api/api-concepts.md:700]: moved

**Review by kevindelgado — COMMENTED** [review]
all feedback addressed, ptal @sftim

**Comment by kevindelgado** [comment]
Build passes now, ptal @sftim

**Comment by sftim** [comment]
@kevindelgado you set a hold on this PR in https://github.com/kubernetes/website/pull/32357#issuecomment-1089020163
Is that still needed?

**Comment by apelisse** [comment]
/unhold

Not needed anymore. Kevin is out for a while.

**Pull request merged**
